File: 21-merge-two-sorted-lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
/* https://leetcode.com/problems/merge-two-sorted-lists/description/?envType=problem-list-v2&envId=linked-list */
function mergeTwoLists(
  list1: ListNode | null,
  list2: ListNode | null,
): ListNode | null {
  let head: ListNode = new ListNode(0);
  let current: ListNode = head;

  while (list1 && list2) {
    const lowestNode = list1.val < list2.val ? list1 : list2;
    current.next = lowestNode;
    current = current.next;
    if (list1.val < list2.val) {
      list1 = list1.next;
    } else {
      list2 = list2.next;
    }
  }

  while (list1) {
    current.next = list1;
    current = current.next;
    list1 = list1.next;
  }

  while (list2) {
    current.next = list2;
    current = current.next;
    list2 = list2.next;
  }

  return head?.next;
}

File: 21-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# 21 Merge Two Sorted Lists
# https://leetcode.com/problems/merge-two-sorted-lists/

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:   
        mergedList = ListNode(None)

        curr = mergedList
        while (list1 and list2):
            if(list1.val > list2.val):
                curr.next = list2
                list2 = list2.next
                print('list 2')
                print(list2)
            else: 
                curr.next = list1
                list1 = list1.next
                print('list1')
                print(list1)
            curr = curr.next
        
        if (list1):
            curr.next = list1
        if (list2):
            curr.next = list2

        return mergedList.next